Security/Guard/Authenticator/OAuth2Authenticator.php

```diff
@@ -0,0 +1,143 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Trikoder\Bundle\OAuth2Bundle\Security\Guard\Authenticator;
+
+use League\OAuth2\Server\Exception\OAuthServerException;
+use League\OAuth2\Server\ResourceServer;
+use Symfony\Bridge\PsrHttpMessage\HttpFoundationFactoryInterface;
+use Symfony\Bridge\PsrHttpMessage\HttpMessageFactoryInterface;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\HttpException;
+use Symfony\Component\HttpKernel\Exception\UnauthorizedHttpException;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+use Symfony\Component\Security\Guard\AuthenticatorInterface;
+use Trikoder\Bundle\OAuth2Bundle\Security\Authentication\Token\OAuth2Token;
+use Trikoder\Bundle\OAuth2Bundle\Security\Authentication\Token\OAuth2TokenFactory;
+use Trikoder\Bundle\OAuth2Bundle\Security\Exception\InsufficientScopesException;
+use Trikoder\Bundle\OAuth2Bundle\Security\Exception\Oauth2AuthenticationFailedException;
+use Trikoder\Bundle\OAuth2Bundle\Security\User\NullUser;
+
+/**
+ * @author Yonel Ceruto <yonelceruto@gmail.com>
+ * @author Antonio J. García Lagar <aj@garcialagar.es>
+ */
+final class OAuth2Authenticator implements AuthenticatorInterface
+{
+    private $httpMessageFactory;
+    private $resourceServer;
+    private $oauth2TokenFactory;
+    private $httpFoundationFactory;
+    private $psr7Request;
+
+    public function __construct(HttpMessageFactoryInterface $httpMessageFactory, ResourceServer $resourceServer, OAuth2TokenFactory $oauth2TokenFactory, HttpFoundationFactoryInterface $httpFoundationFactory)
+    {
+        $this->httpMessageFactory = $httpMessageFactory;
+        $this->resourceServer = $resourceServer;
+        $this->oauth2TokenFactory = $oauth2TokenFactory;
+        $this->httpFoundationFactory = $httpFoundationFactory;
+    }
+
+    public function start(Request $request, ?AuthenticationException $authException = null): Response
+    {
+        if ($authException instanceof InsufficientScopesException || $authException instanceof Oauth2AuthenticationFailedException) {
+            $httpException = new HttpException($authException->getCode(), $authException->getMessage(), $authException);
+        } else {
+            $httpException = new UnauthorizedHttpException('Bearer', $authException ? $authException->getMessage() : null, $authException);
+        }
+
+        return new JsonResponse(['code' => $httpException->getCode(), 'message' => $httpException->getMessage()], $httpException->getStatusCode(), $httpException->getHeaders());
```

Comment:
@yceruto I'm not sure why you need a JsonResponse here. I think that the response type should depend on the request format? WDYT?

Reply:
Yes, it's only my end. For general purpose we have to take "the request format" into account.

In Symfony, we use the `ErrorHandler/ErrorRenderer/SerializerErrorRenderer` (introduced in 4.4) to return the right error response according to the request format. Depending on the bundle's min requirements, we could use it.

Reply:
I've tested it and is doable, but it introduces a new dependency (`symfony/serializer`) and makes the guard authentication behaviour different from the old firewall.

I think that guard authenticator should mimic the firewall behaviour. Maybe we should introduce `ErrorHandler/ErrorRenderer/SerializerErrorRenderer` in a different PR